doc/nvgre_tunnel/nvgre_tunnel.md

```diff
@@ -0,0 +1,446 @@
+  - YANG model should be created in order to auto-generate CLI by using the [SONiC CLI Auto-generation tool](https://github.com/Azure/SONiC/blob/master/doc/cli_auto_generation/cli_auto_generation.md).
+- Phase #2
+  - CLI for NVGRE tunnel
+
+### Architecture Design 
+
+The new tables will be added to Config DB. Unless otherwise stated, the attributes are mandatory.
+
+#### NVGRE configDB table
+
+```
+NVGRE_TUNNEL|{{tunnel_name}} 
+    "src_ip": {{ip_address}} 
+
+NVGRE_TUNNEL_MAP|{{tunnel_name}}|{{tunnel_map}}
+    "vsid": {{vsid_id}}
+    "vlan": {{vlan_id}}
+```
+
+#### ConfigDB schemas
+
+```
+; Defines schema for NVGRE Tunnel configuration attributes
+key                                   = NVGRE_TUNNEL:name             ; NVGRE tunnel configuration
+; field                               = value
+SRC_IP                                = ipv4                          ; IPv4 source address
+
+;value annotations
+ipv4          = dec-octet "." dec-octet "." dec-octet "." dec-octet
+dec-octet     = DIGIT                     ; 0-9  
+                  / %x31-39 DIGIT         ; 10-99  
+                  / "1" 2DIGIT            ; 100-199  
+                  / "2" %x30-34 DIGIT     ; 200-249
+```
+
+```
+; Defines schema for NVGRE Tunnel map configuration attributes
+key                                   = NVGRE_TUNNEL:tunnel_name:name ; NVGRE tunnel configuration
+; field                               = value
+VSID                                  = DIGITS                        ; 1 to 16 million values
+VLAN                                  = 1\*4DIGIT                     ; 1 to 4094 Vlan id
+```
+
+#### Orchestration agent
+
+The following orchestration agents will be added or modified. The flow diagrams are captured in a later section.
+
+##### Figure 1. Orchestration agents
+
+<p align=center>
+<img src="images/nvgre_orch.svg" alt="Figure 1. Orchestration agents">
+</p>
+
+#### NvgreOrch
+
+`nvgreorch` - it is an orchestration agent that handles the configuration requests directly from ConfigDB. The `nvgreorch` is responsible for creating the tunnel and attaches encap and decap mappers. Both VLAN and Bridge tunnel maps are created for NVGRE tunnel by default, so it would be possible to apply both VLAN/VSID and Bridge/VSID map entries.
+
+#### Orchdaemon
+
+`orchdaemon` - it is the main orchestration agent, which handles all Redis DB's updates, then calls appropriate orchagent, the new `nvgreorch` should be registered inside an `orchdaemon`.
+
+### High-Level Design 
+
+The following sub-modules will be modified:
+* [sonic-swss](https://github.com/Azure/sonic-swss) - will be extended with the new orchestration agent for NVGRE.
+* [sonic-swss-common](https://github.com/Azure/sonic-swss-common) - will be extended with the new tables for ConfigDB.
+* [sonic-utilities](https://github.com/Azure/sonic-utilities) - will be extened with the new CLI.
+
+#### Sequence diagrams
+
+##### Figure 2. NVGRE Tunnel creation flow
+
```

Comment:
I think in case tunnel exists we shouldn't raise runtime error but just print a message and continue working since it is not a critical problem.
Maybe even change severity to "Warning" instead, please check in other orchagents to undestand what is commonly used in SONiC.

Runtime error should be raised in orchagent context but not in SAI, please fix.

Please do the same according to above comments for other diagrams.